File: code/server/src/phylo_lens_server/domain/ancillary.py
"""Domain vocabulary: observations, their summaries, and represented isolates."""

from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, Field

AncillaryValue = str | float | bool | None
AncillaryData = dict[str, AncillaryValue]
PositiveCount = Annotated[int, Field(strict=True, ge=1)]


class AncillaryType(StrEnum):
    STRING = "string"
    NUMBER = "number"
    BOOLEAN = "boolean"
    NULL = "null"


class AncillaryField(BaseModel):
    key: str = Field(min_length=1)
    type: AncillaryType


class AncillarySummary(BaseModel):
    values: AncillaryData = Field(default_factory=dict)
    category_counts: dict[str, dict[str, PositiveCount]] = Field(default_factory=dict)


class ProfileSummary(BaseModel):
    # None means unavailable, not zero. LoD nodes sum the represented isolates.
    isolate_count: PositiveCount | None = None


class NodeAnnotations(BaseModel):
    ancillary_data: AncillaryData = Field(default_factory=dict)
    ancillary_summary: AncillarySummary = Field(default_factory=AncillarySummary)
    profile_summary: ProfileSummary = Field(default_factory=ProfileSummary)


class AncillaryObservation(BaseModel):
    """One joint ancillary row shared by count represented isolates."""

    values: AncillaryData = Field(default_factory=dict)
    count: PositiveCount

File: code/client/src/contracts/ancillary.ts
/** User-supplied observations; computed profile counts do not belong here. */
export type AncillaryValue = string | number | boolean | null;
export type AncillaryData = Record<string, AncillaryValue>;
export type AncillaryType = "string" | "number" | "boolean" | "null";
export interface AncillaryField {
  key: string;
  type: AncillaryType;
}
export interface AncillarySummary {
  values: AncillaryData;
  categoryCounts: Record<string, Record<string, number>>;
}
export interface ProfileSummary {
  isolateCount?: number;
}
export interface NodeAnnotations {
  ancillaryData: AncillaryData;
  ancillarySummary: AncillarySummary;
  profileSummary: ProfileSummary;
}
export interface Isolate {
  id: string;
  ancillaryData: AncillaryData;
}
export interface AncillaryTableInput {
  content: string;
  join_column: string;
  format: "auto" | "csv" | "tsv";
}

/** A joint observation shared by count represented isolates. */
export interface AncillaryObservation {
  values: AncillaryData;
  count: number;
}
